src/hotspot/share/gc/shared/partialArrayTaskStats.hpp

```diff
@@ -0,0 +1,107 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_GC_SHARED_PARTIALARRAYTASKSTATS_HPP
+#define SHARE_GC_SHARED_PARTIALARRAYTASKSTATS_HPP
+
+#include "logging/logHandle.hpp"
+#include "logging/logStream.hpp"
+#include "utilities/globalDefinitions.hpp"
+
+#if TASKQUEUE_STATS
+
+class outputStream;
+
+// Repository for collecting and reporting statistics about partial array task
+// processing.  Not thread-safe; each processing thread should have its own
+// stats object.
+class PartialArrayTaskStats {
+  size_t _split;
+  size_t _pushed;
+  size_t _stolen;
+  size_t _processed;
+
+  static LogTargetHandle log_target();
+  static bool is_log_enabled();
+  static void print_header(outputStream* s, const char* title);
+  void print_values(outputStream* s, uint id) const;
+  void print_total(outputStream* s) const;
+  void print_values_impl(outputStream* s) const;
+
+  void accumulate(const PartialArrayTaskStats& stats);
+
+public:
+  // All counters are initially zero.
+  PartialArrayTaskStats();
+
+  // Trivially copied and destroyed.
+
+  // Number of arrays split into partial array tasks.
+  size_t split() const { return _split; }
+
+  // Number of partial array tasks pushed onto a queue.
+  size_t pushed() const { return _pushed; }
+
+  // Number of partial array tasks stolen from some other queue.
+  size_t stolen() const { return _stolen; }
+
+  // Number of partial array tasks processed.
+  size_t processed() const { return _processed; }
+
+  void inc_split() { _split += 1; }
+  void inc_pushed(size_t n) { _pushed += n; }
+  void inc_stolen() { _stolen += 1; }
+  void inc_processed() { _processed += 1; }
+
+  // Set all counters to zero.
+  void reset();
+
+  // Log a table of statistics, if logging is enabled (gc+task+stats=trace).
+  //
+  // num_stats: The number of stats objects to include in the table, one row
+  // for each.
+  //
+  // access: A function taking a uint value < num_stats, and returning a
+  // pointer to the corresponding stats object.
+  //
+  // title: A string title for the table.
+  template<typename StatsAccess>
+  static void log_set(uint num_stats, StatsAccess access, const char* title) {
```

Comment:
Going through all its call sites, I believe `print_stats` is more readable.

Reply:
The name log_set was chosen to suggest that it does "UL logging", and to
indicate that it is for dealing with a set of stats objects.  I think
print_stats loses both of those cues and is less clear because of that.

Reply:
Why is "set" more than important than "stats" in "set of stats objects"? If "UL logging" is critical, "log_stats" would be better. When I first read this name, I thought it's related to "set" as in "getter/setter" of log...

Reply:
"stats" is redundent here.  Recall this is a static function. A client call is
going to look like `PartialArrayTaskStats::log_set(...)`, so it's already
obvious it's related to "stats" at the call site.

A value assigning function would have a "set_" prefix.  Using a "_set" suffix
for that would be really weird and non-idiomatic (and a reader would be quite
right to complain about such).

Reply:
I don't feel that the redundancy here is bad, since the first two args are tied to "stats". OTOH, I find the trailing "set" super confusing. This function is to log/print multiple stats, and the most intuitive choice would have been "log/print" + "stats", because it directly communicates the action being performed (logging stats).

Emphasizing the collective noun instead of the actual noun seems odd. YMMV.